src/backend/core/src/main/scala/com/lightbend/statefulserverless/StatsCollector.scala

```diff
@@ -0,0 +1,206 @@
+      DestinationRevLabel -> settings.revisionName,
+      DestinationPodLabel -> settings.podName
+    )
+    labels.map(map.apply)
+  }
+
+  private val operationsPerSecondGauge = gauges.OperationsPerSecond.labels(labelValues: _*)
+  private val proxiedOperationsPerSecondGauge = gauges.ProxiedOperationsPerSecond.labels(labelValues:_ *)
+  private val averageConcurrentRequestsGauge = gauges.AverageConcurrentRequests.labels(labelValues:_ *)
+  private val averageProxiedConcurrentRequestsGauge = gauges.AverageProxiedConcurrentRequests.labels(labelValues: _*)
+
+
+  private var requestCount: Int = 0
+  private var proxiedCount: Int = 0
+  private var concurrency: Int = 0
+  private var proxiedConcurrency: Int = 0
+
+  private var lastChangedNanos: Long = System.nanoTime()
+  // Mutable maps for performance because these will be frequently updated
+  // Possible optimisation: use arrays for concurrency levels below a certain threshold
+  private val timeNanosOnConcurrency: mutable.Map[Int, Long] = new mutable.HashMap().withDefaultValue(0)
+  private val timeNanosOnProxiedConcurrency: mutable.Map[Int, Long] = new mutable.HashMap().withDefaultValue(0)
+
+  private var lastReportedMillis = System.currentTimeMillis()
+
+  // Report every second - this
+  timers.startPeriodicTimer("tick", Tick, settings.reportPeriod)
+
+  private def updateState(): Unit = {
+    val currentNanos = System.nanoTime()
+    val sinceLastNanos = currentNanos - lastChangedNanos
+    timeNanosOnConcurrency.update(concurrency, timeNanosOnConcurrency(concurrency) + sinceLastNanos)
+    timeNanosOnProxiedConcurrency.update(proxiedConcurrency, timeNanosOnProxiedConcurrency(proxiedConcurrency) + sinceLastNanos)
+    lastChangedNanos = currentNanos
+  }
+
+
+  private def weightedAverage(times: mutable.Map[Int, Long]): Double = {
+
+    // This replicates this go code:
+    //	var totalTimeUsed time.Duration
+    //	for _, val := range times {
+    //		totalTimeUsed += val
+    //	}
+    // 	avg := 0.0
+    //	if totalTimeUsed > 0 {
+    //		sum := 0.0
+    //		for c, val := range times {
+    //			sum += float64(c) * val.Seconds()
+    //		}
+    //		avg = sum / totalTimeUsed.Seconds()
+    //	}
+    //	return avg
+    def toSeconds(value: Long): Double = value.asInstanceOf[Double] / SecondInNanos
+
+    val totalTimeUsed = times.values.sum
+    if (totalTimeUsed > 0) {
+      times.map {
+        case (c, value) => c * toSeconds(value)
+      }.sum / toSeconds(totalTimeUsed)
+    } else 0.0
+  }
+
+  override def receive: Receive = {
+    case CommandSent(proxied) =>
+      updateState()
+      requestCount += 1
+      concurrency += 1
+      if (proxied) {
+        proxiedCount += 1
+        proxiedConcurrency += 1
+      }
+    case ReplyReceived(proxied) =>
+      updateState()
+      concurrency -= 1
+      if (proxied) {
+        proxiedConcurrency -= 1
+      }
+    case Tick =>
+      val currentTime = System.currentTimeMillis()
```

Comment:
currentTimeMillis is not a really good source of time, since it can travel backwards etc.